File: 0790-domino-and-tromino-tiling/0790-domino-and-tromino-tiling.cpp
const int MOD = 1e9 + 7;

class Solution {
public:
    int numTilings(int n) {
        unsigned int n0 = 1, n1 = 1,n2 = 2, n3;
        if (n < 2) {
            return 1;
        }
        if (n == 2) {
            return 2;
        }
        for (int i = 3; i <= n; i++) {
            n3 = ((2 * n2) + n0) % MOD;
            n0 = n1;
            n1 = n2;
            n2 = n3;
        }
        return n3;
    }
};

File: 0790-domino-and-tromino-tiling/0790-domino-and-tromino-tiling.java
class Solution {
    
    static long[] dp;
    
    static {
        int MOD = 1000000007;
        dp = new long[1001];
        dp[1] = 1;
        dp[2] = 2;
        dp[3] = 5;
        for (int i = 4; i <= 1000; i++) {
            dp[i] = 2 * dp[i - 1] + dp[i - 3]; 
            dp[i] %= MOD;
        }
    }
    
    public int numTilings(int n) {
        return (int) dp[n];
    }
}